### PR TITLE
Add ability to fallback to chgrp remote_tmp and its files.

### LINKED ISSUE #68284: Add Group Assignment and Permissions to Shell Plugin for remote_tmp
##### SUMMARY

Add the ability to assign a group and group permissions to remote_tmp through the shell plugin

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME

sh – POSIX shell (/bin/sh)

##### ADDITIONAL INFORMATION

```ini
[defaults]
remote_tmp_group=groupname
remote_tmp_group_permissions=true
```

Alternatively, detect that `remote_tmp_group` is defined and if so, automatically apply `770` permissions to the remote_tmp directory.

Please also include an inventory variable `ansible_remote_tmp_group` similar to `ansible_remote_tmp`.

### PR DESCRIPTION
**Pull request by relrod**
##### SUMMARY

Adds a fallback for chgrp-ing the tmpdir we create if setfacl and chown both fail.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME

`plugins/action/__init__.py`

### CODE REVIEW

File: lib/ansible/config/base.yml [hunk 1]
```diff
@@ -328,6 +328,16 @@ COLOR_WARN:
   env: [{name: ANSIBLE_COLOR_WARN}]
   ini:
   - {key: warn, section: colors}
+COMMON_REMOTE_GROUP:
```
Comment (bcoca) [inline lib/ansible/config/base.yml]: don't do this in base, do it in shell plugin so you can also allow per host vars

Comment (bcoca) [reply lib/ansible/config/base.yml]: missing version_added

Comment (relrod) [reply lib/ansible/config/base.yml]: Added `version_added`. Curious what others (@sivel, @nitzmahone, @abadger) think about where this should be.

I actually did originally have this in the shell plugin config, but moved it out to base for two reasons:

* This is a setting needed with every shell; it shouldn't be up to a shell plugin to decide if this gets disabled. Changing to a group is a platform abstraction, not a shell abstraction, and platform abstractions are currently managed by the connection plugin (see `_IS_WINDOWS`), not the shell.
* This only belongs with and affects `fixup_perms2()` which is in action, not shell, and therefore a different subsystem.

**Comment by relrod** [comment]
@acozine @samccann I've tried updating the `become` user guide as part of this change, but fear that my changes have made the docs too implementation detail-ey. Could you review and let me know what you think and/or how to make them more clear?

**Comment by relrod** [comment]
Adding @mattclay as a reviewer too because these tests that I added are the bane of my existence and I probably messed them up somehow.

File: lib/ansible/config/base.yml (hunk 1 above)
Comment (bcoca) [reply lib/ansible/config/base.yml]: not all platforms work the same, some like windows always have acls, others like networking os make all of this irrelevant, others even allow direct assignment to user w/o having to resort to groups.

Comment (relrod) [reply lib/ansible/config/base.yml]: Ok, moving it to shell.

File: docs/docsite/rst/user_guide/become.rst [hunk 2]
```diff
@@ -130,28 +133,57 @@ that only allow reading by the user and root, or only allow reading by the unpri
...
 However, when both the connection user and the ``become_user`` are unprivileged,
-the module file is written as the user that Ansible connects as, but the file needs to
-be readable by the user Ansible is set to ``become``. In this case, Ansible makes
-the module file world-readable for the duration of the Ansible module execution.
-Once the module is done executing, Ansible deletes the temporary file.
+the module file is written as the user that Ansible connects as (the
+``remote_user``), but the file needs to be readable by the user Ansible is set
+to ``become``. Ansible solves this problem in the following way:
+
+First, if :command:`setfacl` is installed and available in the remote ``PATH``,
```
Comment (bcoca) [inline docs/docsite/rst/user_guide/become.rst:141]: i would note 'this depends on the platform, but for POSIX systems this is how it works' or similar phrasing

File: docs/docsite/rst/user_guide/become.rst [hunk 3]
```diff
@@ -130,28 +133,57 @@ that only allow reading by the user and root, or only allow reading by the unpri
...
+and the temporary directory on the remote host is mounted with POSIX.1e
+filesystem ACL support, Ansible will use POSIX ACLs to share the module file
+with the second unprivileged user.
+
+Next, if POSIX ACLs are **not** available or :command:`setfacl` could not be
+run, Ansible will attempt to change ownership of the module file using
+:command:`chown` for systems which support doing so as an unprivileged user.
+
+If the :command:`chown` fails, Ansible will then check the value of the
+configuration setting ``ansible_common_remote_group``. Many systems will allow a
```
Comment (bcoca) [inline docs/docsite/rst/user_guide/become.rst]: 'new in 2.10', i know you mention it below but most users won't read that far, so it is good to show version the first time a thing is introduced in the docs

File: docs/docsite/rst/user_guide/become.rst [hunk 4]
```diff
@@ -83,6 +83,9 @@ ansible_become_user
 ansible_become_password
     set the privilege escalation password. See :ref:`playbooks_vault` for details on how to avoid having secrets in plain text
 
+ansible_common_remote_group
+    determines if Ansible should try to ``chgrp`` its temporary files to a group if ``setfacl`` and ``chown`` both fail. See `Risks of becoming an unprivileged user`_ for more information.
+
```
Comment (bcoca) [inline docs/docsite/rst/user_guide/become.rst:88]: add version added tag

File: docs/docsite/rst/user_guide/become.rst [hunk 5]
```diff
@@ -72,7 +72,7 @@ Become connection variables
 You can define different ``become`` options for each managed node or group. You can define these variables in inventory or use them as normal variables.
 
 ansible_become
-    equivalent of the become directive, decides if privilege escalation is used or not.
+    equivalent of the ``become`` directive, decides if privilege escalation is used or not.
```
Comment (bcoca) [inline docs/docsite/rst/user_guide/become.rst]: not an issue with the change, but the original phrasing was already 'wrongish' .. its not equivalent, it overrides

File: lib/ansible/plugins/action/__init__.py [hunk 6]
```diff
@@ -520,12 +535,48 @@ def _fixup_perms2(self, remote_paths, remote_user=None, execute=True):
...
+
+                    # Otherwise, we're a normal user. We failed to chown the
+                    # paths to the unprivileged user, but if we have a common
+                    # group with them, we should be able to chown it to that.
+                    #
+                    # Note that we have no way of knowing if this will actually
+                    # work... just because chgrp exits successfully does not
+                    # mean that Ansible will work. We could check if the become
+                    # user is in the group, but this would create an extra
+                    # round trip.
```
Comment (bcoca) [inline lib/ansible/plugins/action/__init__.py:585]: not really, we could append/chain the verification to the end of the same command string, but that is also 'fun' in certain contexts that will misreport, so we are better off just letting the final execution fail for permission issues

**Review by bcoca — CHANGES_REQUESTED** [review]
very minor stuff in docs, code lgtm

**Review by bcoca — APPROVED** [review]

**Pull request merged**